File: cron/runner.py
"""
Cron job runner.
Executes a Python script, captures its stdout as a prompt,
runs claude -p (--resume if session exists), and writes output to outbox.

Usage:
  uv run cron/runner.py --script email_check.py --topic 신건 --user-id 123 --cron-name email-check
"""

import argparse
import json
import sqlite3
import subprocess
import sys
import time
from pathlib import Path

PROJECT_ROOT = Path(__file__).resolve().parent.parent
CRON_DIR = PROJECT_ROOT / "cron"
DATA_DIR = PROJECT_ROOT / "data"
USERS_DIR = DATA_DIR / "users"
SESSIONS_DB = DATA_DIR / "sessions.db"


def user_dir(user_id: str) -> Path:
    return USERS_DIR / user_id


def lock_dir(user_id: str) -> Path:
    return user_dir(user_id) / "active-queries"


def cron_lock_dir(user_id: str) -> Path:
    return user_dir(user_id) / "cron-locks"


def outbox_dir(user_id: str) -> Path:
    return user_dir(user_id) / "cron-outbox"


CLAUDE_BIN = Path.home() / ".local" / "bin" / "claude"


def get_cron_session_id(server_name: str, topic: str) -> str | None:
    """Look up cron-dedicated session ID from SQLite."""
    try:
        con = sqlite3.connect(str(SESSIONS_DB), timeout=5)
        row = con.execute(
            "SELECT cron_session_id FROM topics WHERE server_name = ? AND name = ?",
            (server_name, topic),
        ).fetchone()
        con.close()
        return row[0] if row and row[0] else None
    except Exception:
        return None


def set_cron_session_id(server_name: str, topic: str, session_id: str):
    """Save cron session ID to SQLite."""
    try:
        con = sqlite3.connect(str(SESSIONS_DB), timeout=5)
        con.execute(
            "UPDATE topics SET cron_session_id = ? WHERE server_name = ? AND name = ?",
            (session_id, server_name, topic),
        )
        con.commit()
        con.close()
    except Exception as e:
        print(f"[runner] Failed to save cron session ID: {e}", file=sys.stderr)


def is_locked(user_id: str, topic: str) -> bool:
    """Check if there's an active query on this topic."""
    state_file = lock_dir(user_id) / f"{topic}.json"
    if not state_file.exists():
        return False
    # Stale lock check (older than 10 minutes)
    try:
        state = json.loads(state_file.read_text().strip())
        from datetime import datetime, timezone
        since = datetime.fromisoformat(state["since"]).replace(tzinfo=timezone.utc)
        elapsed = time.time() - since.timestamp()
        if elapsed > 600:
            state_file.unlink(missing_ok=True)
            return False
    except Exception:
        pass
    return True


def wait_for_unlock(user_id: str, topic: str, timeout: int = 300):
    """Wait until the topic is unlocked, polling every 5 seconds."""
    start = time.time()
    while is_locked(user_id, topic):
        if time.time() - start > timeout:
            print(f"[runner] Timeout waiting for unlock: {user_id}-{topic}", file=sys.stderr)
            return False
        time.sleep(5)
    return True


def write_outbox(user_id: str, topic: str, cron_name: str, message: str, files: list[str] | None = None, session_texts: list[str] | None = None, new_cron_session_id: str | None = None):
    """Write result to outbox for bot.ts to pick up."""
    out_dir = outbox_dir(user_id)
    out_dir.mkdir(parents=True, exist_ok=True)
    outbox_file = out_dir / "pending.jsonl"
    actual_message = "\n\n".join(session_texts) if session_texts else message
    entry = {
        "userId": user_id,
        "topic": topic,
        "cronName": cron_name,
        "message": actual_message,
        "timestamp": time.strftime("%Y-%m-%dT%H:%M:%S"),
    }
    if files:
        entry["files"] = files
    if new_cron_session_id:
        entry["newCronSessionId"] = new_cron_session_id
    with open(outbox_file, "a") as f:
        f.write(json.dumps(entry, ensure_ascii=False) + "\n")


def main():
    parser = argparse.ArgumentParser(description="Cron job runner")
    parser.add_argument("--script", required=True, help="Python script in cron/ to run")
    parser.add_argument("--topic", required=True, help="Telegram topic name")
    parser.add_argument("--user-id", required=True, help="User ID")
    parser.add_argument("--server-name", required=True, help="Server (bot) name for topic lookups")
    parser.add_argument("--cron-name", default="unknown", help="Cron job name")
    args = parser.parse_args()

    script_path = CRON_DIR / args.script
    if not script_path.exists():
        print(f"[runner] Script not found: {script_path}", file=sys.stderr)
        sys.exit(1)

    # 0. Skip if already running (prevent overlap)
    cl_dir = cron_lock_dir(args.user_id)
    cl_dir.mkdir(parents=True, exist_ok=True)
    cron_lock = cl_dir / f"{args.cron_name}.lock"
    if cron_lock.exists():
        # Stale lock check (older than 30 minutes)
        try:
            ts = float(cron_lock.read_text().strip())
            if time.time() - ts < 1800:
                print(f"[runner] Skipping: {args.cron_name} is already running", file=sys.stderr)
                sys.exit(0)
        except Exception:
            pass
    cron_lock.write_text(str(time.time()))

    try:
        _run_job(args, script_path)
    finally:
        cron_lock.unlink(missing_ok=True)


def _run_job(args, script_path):
    # 1. Run the script to get prompt
    try:
        result = subprocess.run(
            [sys.executable, str(script_path)],
            capture_output=True,
            text=True,
            timeout=3600,
        )
        prompt = result.stdout.strip()
        if not prompt:
            print(f"[runner] Script produced no output (normal - nothing to report)", file=sys.stderr)
            sys.exit(0)
    except subprocess.TimeoutExpired:
        print(f"[runner] Script timed out", file=sys.stderr)
        sys.exit(1)
    except Exception as e:
        print(f"[runner] Script error: {e}", file=sys.stderr)
        sys.exit(1)

    # 2. Wait for active query to finish
    if not wait_for_unlock(args.user_id, args.topic):
        write_outbox(args.user_id, args.topic, args.cron_name,
                     f"[cron skipped] Topic '{args.topic}' was busy for too long.")
        sys.exit(1)

    # 3. Run claude -p with stream-json (to capture tool_use events for file sending)
    cron_session_id = get_cron_session_id(args.server_name, args.topic)

    # MCP config for send-file and send-text
    send_file_server = str(PROJECT_ROOT / "src" / "mcp" / "send-file-server.ts")
    send_text_server = str(PROJECT_ROOT / "src" / "mcp" / "send-text-server.ts")
    mcp_config = json.dumps({
        "mcpServers": {
            "send-file": {
                "command": "bun",
                "args": ["run", send_file_server, f"--user-id={args.user_id}"],
            },
            "send-text": {
                "command": "bun",
                "args": ["run", send_text_server],
            },
        }
    })

    claude_cmd = [
        str(CLAUDE_BIN), "-p", prompt,
        "--output-format", "stream-json",
        "--verbose",
        "--dangerously-skip-permissions",
        "--mcp-config", mcp_config,
    ]
    if cron_session_id:
        claude_cmd.extend(["--resume", cron_session_id])

    response = ""
    files: list[str] = []
    session_texts: list[str] = []
    new_session_id: str | None = None

    try:
        proc = subprocess.Popen(
            claude_cmd,
            stdout=subprocess.PIPE,
            stderr=subprocess.PIPE,
            text=True,
        )
        assert proc.stdout is not None

        for line in proc.stdout:
            line = line.strip()
            if not line:
                continue
            try:
                event = json.loads(line)
            except json.JSONDecodeError:
                continue

            event_type = event.get("type", "")

            # Extract file paths from tool_use events (send_file / send_files)
            if event_type == "assistant":
                for block in event.get("message", {}).get("content", []):
                    if block.get("type") != "tool_use":
                        continue
                    tool_name = block.get("name", "")
                    tool_input = block.get("input", {})
                    if tool_name.endswith("send_file"):
                        fp = tool_input.get("file_path", "")
                        if fp:
                            files.append(fp)
                    elif tool_name.endswith("send_files"):
                        for fp in tool_input.get("file_paths", []):
                            if fp:
                                files.append(fp)
                    elif tool_name.endswith("send_text"):
                        text = tool_input.get("text", "")
                        if text:
                            session_texts.append(text)

            # Extract result and session_id from the final result event
            elif event_type == "result":
                response = event.get("result", "")
                new_session_id = event.get("session_id")

        proc.wait(timeout=60)

    except subprocess.TimeoutExpired:
        if proc:
            proc.kill()
        response = "[cron error] Claude CLI timed out"
    except Exception as e:
        response = f"[cron error] {e}"

    if not response:
        response = "(empty response)"

    # Deduplicate files while preserving order
    seen: set[str] = set()
    unique_files = []
    for f in files:
        if f not in seen:
            seen.add(f)
            unique_files.append(f)

    write_outbox(args.user_id, args.topic, args.cron_name, response, unique_files or None, session_texts or None, new_session_id or None)
    print(f"[runner] Done. Output written to outbox for topic '{args.topic}'" +
          (f" ({len(unique_files)} files: {unique_files})" if unique_files else " (no files)") +
          (f" ({len(session_texts)} session texts merged)" if session_texts else ""))


if __name__ == "__main__":
    main()

File: cron/runner.ts
#!/usr/bin/env bun
/**
 * Cron job runner (TypeScript).
 *
 * Executes a Python script, captures its stdout as a prompt,
 * runs runAgent() (with --resume if a cron session exists), and writes
 * the result to the user's cron outbox for bot.ts to pick up.
 *
 * Usage:
 *   bun run cron/runner.ts --script email_check.py --topic "[srv] 신건" --user-id 123 --cron-name email-check
 */

import { Database } from "bun:sqlite";
import {
  appendFileSync,
  existsSync,
  mkdirSync,
  readFileSync,
  unlinkSync,
  writeFileSync,
} from "node:fs";
import { homedir } from "node:os";
import { join } from "node:path";
import { runAgent } from "@/core/agents";
import { buildTopicSystemPrompt, PROJECT_ROOT, SERVER_NAME, SESSIONS_DB, USERS_LOG_DIR } from "@/core/config";
import { SESSION_STALE_SEC } from "@/mcp/cron/shared";

// --- CLI args ---
function parseArg(name: string, required = false): string {
  const flag = `--${name}`;
  const eqMatch = process.argv.find((a) => a.startsWith(`${flag}=`));
  if (eqMatch) return eqMatch.slice(flag.length + 1);
  const idx = process.argv.indexOf(flag);
  const val = idx !== -1 ? process.argv[idx + 1] : undefined;
  if (required && !val) {
    console.error(`[runner] Missing required argument: --${name}`);
    process.exit(1);
  }
  return val ?? "";
}

const scriptName = parseArg("script", true);
const topicName = parseArg("topic", true);
const userId = parseArg("user-id", true);
const cronName = parseArg("cron-name") || "unknown";

// --- Paths ---
const CRON_DIR = join(PROJECT_ROOT, "cron");
const userDir = join(USERS_LOG_DIR, userId);
const lockDir = join(userDir, "active-queries");
const cronLockDir = join(userDir, "cron-locks");
const sessionLockDir = join(cronLockDir, topicName); // topic-scoped: prevents cross-topic blocking
const outboxDir = join(userDir, "cron-outbox");
const cronRunsDir = join(userDir, "cron-runs");
const scriptPath = join(CRON_DIR, scriptName);

// --- DB helpers ---
interface TopicDbRow {
  cron_session_id: string | null;
  cwd: string | null;
  description: string | null;
  model: string | null;
  effort: string | null;
}

function readTopicRow(): TopicDbRow | null {
  if (!existsSync(SESSIONS_DB)) return null;
  try {
    const db = new Database(SESSIONS_DB, { readonly: true });
    db.exec("PRAGMA busy_timeout = 5000");
    const row = db
      .query<TopicDbRow, [string, string]>(
        "SELECT cron_session_id, cwd, description, model, effort FROM topics WHERE server_name = ? AND name = ?",
      )
      .get(SERVER_NAME, topicName);
    db.close();
    return row;
  } catch {
    return null;
  }
}

function setCronSessionId(sessionId: string | null): void {
  try {
    const db = new Database(SESSIONS_DB);
    db.exec("PRAGMA busy_timeout = 5000");
    db.query("UPDATE topics SET cron_session_id = ? WHERE server_name = ? AND name = ?").run(
      sessionId,
      SERVER_NAME,
      topicName,
    );
    db.close();
  } catch (e) {
    console.error(`[runner] Failed to save cron session ID: ${e}`);
  }
}

// --- Run counter (reset session every N runs to keep context bounded) ---
const CRON_SESSION_RESET_EVERY = 5;
const counterFile = join(cronLockDir, `${cronName}.count`);

function getRunCount(): number {
  try {
    return Number.parseInt(readFileSync(counterFile, "utf-8").trim(), 10) || 0;
  } catch {
    return 0;
  }
}

function incrementRunCount(): number {
  mkdirSync(cronLockDir, { recursive: true });
  const next = getRunCount() + 1;
  writeFileSync(counterFile, String(next));
  return next;
}

// --- Run history (last 20 runs) ---
const RUN_HISTORY_MAX = 20;

function appendRunHistory(entry: {
  at: string;
  exit_code: number;
  duration_ms: number;
  output_preview: string;
}): void {
  mkdirSync(cronRunsDir, { recursive: true });
  const runsFile = join(cronRunsDir, `${cronName}.jsonl`);
  if (existsSync(runsFile)) {
    try {
      const lines = readFileSync(runsFile, "utf-8").trim().split("\n").filter(Boolean);
      if (lines.length >= RUN_HISTORY_MAX) {
        writeFileSync(runsFile, `${lines.slice(-(RUN_HISTORY_MAX - 1)).join("\n")}\n`);
      }
    } catch {}
  }
  appendFileSync(runsFile, `${JSON.stringify(entry)}\n`);
}

// --- Session lock (topic-scoped: O_EXCL-atomic; steal if stale) ---
const sessionLockFile = join(sessionLockDir, "_session.lock");

function isSessionLockStale(): boolean {
  try {
    const ts = Number(readFileSync(sessionLockFile, "utf-8").trim());
    return Number.isNaN(ts) || Date.now() / 1000 - ts > SESSION_STALE_SEC;
  } catch {
    return true;
  }
}

/**
 * Wait for session lock to clear, then acquire atomically.
 * `wx` flag gives us O_EXCL so two concurrent runners can't both "acquire" after
 * reading an empty slot. On EEXIST we check staleness (steal if expired) and retry.
 */
async function acquireSessionLockWithWait(timeoutMs = 60_000): Promise<boolean> {
  mkdirSync(sessionLockDir, { recursive: true });
  const start = Date.now();
  while (true) {
    try {
      writeFileSync(sessionLockFile, String(Date.now() / 1000), { flag: "wx" });
      return true;
    } catch (e: unknown) {
      if ((e as NodeJS.ErrnoException)?.code !== "EEXIST") throw e;
      if (isSessionLockStale()) {
        try { unlinkSync(sessionLockFile); } catch {}
        continue;
      }
      if (Date.now() - start > timeoutMs) return false;
      await Bun.sleep(5000);
    }
  }
}

function releaseSessionLock(): void {
  try { if (existsSync(sessionLockFile)) unlinkSync(sessionLockFile); } catch {}
}

// --- Topic lock (bot's active-queries state file) ---
function isTopicLocked(): boolean {
  const stateFile = join(lockDir, `${topicName}.json`);
  if (!existsSync(stateFile)) return false;
  try {
    const state = JSON.parse(readFileSync(stateFile, "utf-8")) as { since: string };
    if (Date.now() - new Date(state.since).getTime() > 600_000) {
      unlinkSync(stateFile);
      return false;
    }
  } catch {}
  return true;
}

async function waitForTopicUnlock(timeoutMs = 300_000): Promise<boolean> {
  const start = Date.now();
  while (isTopicLocked()) {
    if (Date.now() - start > timeoutMs) {
      console.error(`[runner] Timeout waiting for unlock: ${userId}-${topicName}`);
      return false;
    }
    await Bun.sleep(5000);
  }
  return true;
}

// --- Outbox ---
function writeOutbox(message: string, newSessionId?: string): void {
  mkdirSync(outboxDir, { recursive: true });
  const entry: Record<string, unknown> = {
    userId,
    topic: topicName,
    cronName,
    message,
    timestamp: new Date().toISOString().slice(0, 19),
  };
  if (newSessionId) entry.newCronSessionId = newSessionId;
  appendFileSync(join(outboxDir, "pending.jsonl"), `${JSON.stringify(entry)}\n`);
}

// --- Main ---
async function main() {
  if (!existsSync(scriptPath)) {
    console.error(`[runner] Script not found: ${scriptPath}`);
    process.exit(1);
  }

  // Skip-first marker: set by cron-manager on create/restart to suppress pm2's
  // immediate-on-start execution. Consume it once and exit without firing.
  const skipFirstMarker = join(cronLockDir, `${cronName}.skip-first`);
  if (existsSync(skipFirstMarker)) {
    try { unlinkSync(skipFirstMarker); } catch {}
    console.error(`[runner] Suppressing first run after registration (${cronName}); next run will fire at scheduled tick.`);
    process.exit(0);
  }

  // Job lock (prevent same job overlapping)
  mkdirSync(cronLockDir, { recursive: true });
  const jobLock = join(cronLockDir, `${cronName}.lock`);
  if (existsSync(jobLock)) {
    try {
      const ts = Number(readFileSync(jobLock, "utf-8").trim());
      if (!Number.isNaN(ts) && Date.now() / 1000 - ts < SESSION_STALE_SEC) {
        console.error(`[runner] Skipping: ${cronName} is already running`);
        process.exit(0);
      }
    } catch {}
  }
  writeFileSync(jobLock, String(Date.now() / 1000));

  try {
    await runJob();
  } finally {
    try { unlinkSync(jobLock); } catch {}
  }
}

async function runJob() {
  const topicRow = readTopicRow();
  // ~/ 등 tilde 확장 처리 (Bun.spawnSync는 shell expansion 미지원)
  const rawCwd = topicRow?.cwd || homedir();
  const workDir = rawCwd.replace(/^~/, homedir());

  // 1. Run Python script to get prompt
  // uv 절대경로 사용 (pm2 환경에서 ~/.local/bin이 PATH에 없을 수 있음)
  const uvBin = `${homedir()}/.local/bin/uv`;
  const proc = Bun.spawnSync([uvBin, "run", "--project", CRON_DIR, "python", scriptPath], {
    cwd: workDir,
    stderr: "pipe",
  });
  if (proc.exitCode !== 0) {
    const stderr = new TextDecoder().decode(proc.stderr).trim();
    throw new Error(`Script exited with code ${proc.exitCode}: ${stderr}`);
  }
  const prompt = new TextDecoder().decode(proc.stdout).trim();
  if (!prompt) {
    console.error("[runner] Script produced no output (normal - nothing to report)");
    return;
  }

  // 2. Wait for interactive topic to unlock
  if (!(await waitForTopicUnlock())) {
    writeOutbox(`[cron skipped] Topic '${topicName}' was busy for too long.`);
    return;
  }

  // 3. Acquire session lock atomically BEFORE session reset / getCronSessionId —
  //    otherwise a concurrent finisher could overwrite our NULL reset or hand us
  //    a stale session id mid-flight.
  if (!(await acquireSessionLockWithWait())) {
    writeOutbox(`[cron skipped] Cron session busy for too long (${cronName}).`);
    return;
  }

  // 4. Session reset every N runs (bounded context window)
  const runCount = incrementRunCount();
  if (runCount % CRON_SESSION_RESET_EVERY === 0) {
    console.error(`[runner] Run ${runCount}: resetting cron session`);
    setCronSessionId(null);
  }

  const freshRow = readTopicRow();
  const cronSessionId = freshRow?.cron_session_id ?? null;

  let response = "";
  let newSessionId: string | undefined;
  const startedAt = new Date().toISOString();
  const startMs = Date.now();
  let exitCode = 0;

  try {
    for await (const event of runAgent({
      prompt,
      sessionId: cronSessionId,
      cwd: workDir,
      userId,
      session: topicName,
      sessionType: "forum",
      systemPrompt: buildTopicSystemPrompt({ description: freshRow?.description ?? null }),
      ...(freshRow?.model && { model: freshRow.model }),
      ...(freshRow?.effort && { effort: freshRow.effort as "low" | "medium" | "high" | "max" }),
    })) {
      switch (event.type) {
        case "session":
          newSessionId = event.sessionId;
          break;
        case "result":
          response = event.content;
          break;
      }
    }
  } catch (e) {
    response = `[cron error] ${e}`;
    exitCode = 1;
  } finally {
    releaseSessionLock();

    if (!response) response = "(empty response)";
    const durationMs = Date.now() - startMs;

    writeOutbox(response, newSessionId);
    if (newSessionId) setCronSessionId(newSessionId);
    appendRunHistory({
      at: startedAt,
      exit_code: exitCode,
      duration_ms: durationMs,
      output_preview: response.slice(0, 150),
    });

    console.error(`[runner] Done. Output written to outbox for topic '${topicName}'`);
  }
}

main().catch((e) => {
  console.error(`[runner] Fatal error: ${e}`);
  process.exit(1);
});
